src/node_contextify.cc

```diff
@@ -441,40 +373,111 @@ class ContextifyContext {
     bool is_function = value->IsFunction();
 
     if (!is_declared && args.ShouldThrowOnError() && is_contextual_store &&
-    !is_function)
+        !is_function)
       return;
 
     ctx->sandbox()->Set(property, value);
   }
 
 
-  static void GlobalPropertyQueryCallback(
+  static void GlobalPropertyDescriptorCallback(
       Local<Name> property,
-      const PropertyCallbackInfo<Integer>& args) {
+      const PropertyCallbackInfo<Value>& info) {
     ContextifyContext* ctx;
-    ASSIGN_OR_RETURN_UNWRAP(&ctx, args.Data().As<Object>());
+    ASSIGN_OR_RETURN_UNWRAP(&ctx, info.Data().As<Object>());
 
     // Still initializing
     if (ctx->context_.IsEmpty())
       return;
 
     Local<Context> context = ctx->context();
-    Maybe<PropertyAttribute> maybe_prop_attr =
-        ctx->sandbox()->GetRealNamedPropertyAttributes(context, property);
+    v8::Isolate* isolate = context->GetIsolate();
+    HandleScope scope(isolate);
 
-    if (maybe_prop_attr.IsNothing()) {
-      maybe_prop_attr =
-          ctx->global_proxy()->GetRealNamedPropertyAttributes(context,
-                                                              property);
-    }
+    Local<String> key = property->ToString(isolate);
+    Local<Object> sandbox = ctx->sandbox();
 
-    if (maybe_prop_attr.IsJust()) {
-      PropertyAttribute prop_attr = maybe_prop_attr.FromJust();
-      args.GetReturnValue().Set(prop_attr);
+    Maybe<bool> has = sandbox->HasOwnProperty(context, key);
+    Local<Value> descriptor_intercepted = has.IsNothing() ?
+        ctx->global_proxy()->GetOwnPropertyDescriptor(context, key)
+            .ToLocalChecked() :
+        sandbox->GetOwnPropertyDescriptor(context, key).ToLocalChecked();
+
+    // If the property had already been defined on the sandbox or global object,
+    // we intercept, query and return it.
+    // Else, there is no need to intercept with first time definitions.
+    if (!descriptor_intercepted->IsUndefined()) {
```

Comment:
I suppose the conditional isn't needed?

Reply:
Why? If we intercept, it signals that the property was found. I think that's not correct if the property is not present.

Reply:
Undefined is the default return value of API callbacks.  I.e., the return value is going to be the same, with or without the `IsUndefined()` check.

Reply:
`ReturnValue` on the `CallbackArgs` is set to `undefined`, but the call to the interceptor returns `true` or `false` depending on whether the return value was set. This then changes the control flow, see [objects.cc](https://github.com/v8/v8/blob/6f93d59d9204101ee11c3ed9513c7012b0f75169/src/objects.cc#L7673).

But taking a closer look, I think the logic can be substantially simplified: take the value from the sandbox if present, otherwise proceed regularly. What do you think?